source/extensions/filters/udp/udp_proxy/router/router_impl.cc

```diff
@@ -0,0 +1,84 @@
+#include "source/extensions/filters/udp/udp_proxy/router/router_impl.h"
+
+#include "envoy/extensions/filters/udp/udp_proxy/v3/route.pb.h"
+#include "envoy/extensions/filters/udp/udp_proxy/v3/route.pb.validate.h"
+#include "envoy/type/matcher/v3/network_inputs.pb.h"
+
+#include "source/common/common/empty_string.h"
+#include "source/common/network/matching/data_impl.h"
+#include "source/common/network/matching/inputs.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace UdpFilters {
+namespace UdpProxy {
+namespace Router {
+
+Matcher::ActionFactoryCb RouteMatchActionFactory::createActionFactoryCb(
+    const Protobuf::Message& config, RouteActionContext& context,
+    ProtobufMessage::ValidationVisitor& validation_visitor) {
+  const auto& route_config = MessageUtil::downcastAndValidate<
+      const envoy::extensions::filters::udp::udp_proxy::v3::Route&>(config, validation_visitor);
+  const auto& cluster = route_config.cluster();
+
+  // Emplace cluster to context to get all clusters
+  context.entries_.emplace(cluster);
+
+  return [cluster]() { return std::make_unique<RouteMatchAction>(cluster); };
+}
+
+REGISTER_FACTORY(RouteMatchActionFactory, Matcher::ActionFactory<RouteActionContext>);
+
+absl::Status RouteActionValidationVisitor::performDataInputValidation(
+    const Matcher::DataInputFactory<Network::NetworkMatchingData>&, absl::string_view type_url) {
+  static std::string source_ip_input_name = TypeUtil::descriptorFullNameToTypeUrl(
+      envoy::type::matcher::v3::SourceIpMatchInput::descriptor()->full_name());
+  if (type_url == source_ip_input_name) {
+    return absl::OkStatus();
+  }
+
+  return absl::InvalidArgumentError(
+      fmt::format("Route table can only match on source IP, saw {}", type_url));
```

Comment:
Ok, I will remove these data inputs, but I still think the validation is necessary since a HTTP data input is also syntax valid for the UDP router and the validator will keep the gate.

Reply:
The way the factory registrations work is that registrations are unique per data input type, so since you're using the NetworkDataInput it won't share registrations with HttpDataInput registrations

Reply:
https://github.com/envoyproxy/envoy/blob/818419918a6b9e05f9fdb42302c84ed4c784607a/test/extensions/filters/udp/udp_proxy/router_impl_test.cc#L218
Yes, I also think matcher is expected to work that way, but the test shows Envoy will not stop until the validation visitor returns the error. I guess the registration of `HttpDataInput` is shared with `NetworkDataInput`, do I miss something?